src/core/public/http/fetch.ts

```diff
@@ -0,0 +1,154 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import Url from 'url';
+
+import { merge } from '../../utils';
+import { IBasePath, HttpInterceptor, HttpHandler, HttpFetchOptions, IHttpResponse } from './types';
+import { HttpFetchError } from './http_fetch_error';
+import { HttpInterceptController } from './http_intercept_controller';
+import { HttpResponse } from './response';
+import { interceptRequest, interceptResponse } from './intercept';
+import { HttpInterceptHaltError } from './http_intercept_halt_error';
+
+interface Params {
+  basePath: IBasePath;
+  kibanaVersion: string;
+}
+
+const JSON_CONTENT = /^(application\/(json|x-javascript)|text\/(x-)?javascript|x-json)(;.*)?$/;
+const NDJSON_CONTENT = /^(application\/ndjson)(;.*)?$/;
+
+export class FetchService {
+  private readonly interceptors = new Set<HttpInterceptor>();
+
+  constructor(private readonly params: Params) {}
+
+  public intercept(interceptor: HttpInterceptor) {
+    this.interceptors.add(interceptor);
+    return () => this.interceptors.delete(interceptor);
+  }
+
+  public removeAllInterceptors() {
+    this.interceptors.clear();
+  }
+
+  public fetch: HttpHandler = async <TResponseBody>(
+    path: string,
+    options: HttpFetchOptions = {}
+  ) => {
+    const initialRequest = this.createRequest(path, options);
+    const controller = new HttpInterceptController();
+
+    // We wrap the interception in a separate promise to ensure that when
+    // a halt is called we do not resolve or reject, halting handling of the promise.
+    return new Promise<TResponseBody | IHttpResponse<TResponseBody>>(async (resolve, reject) => {
+      try {
+        const interceptedRequest = await interceptRequest(
+          initialRequest,
+          this.interceptors,
+          controller
+        );
+        const initialResponse = this.fetchResponse(interceptedRequest);
+        const interceptedResponse = await interceptResponse(
+          initialResponse,
+          this.interceptors,
+          controller
+        );
+
+        if (options.asResponse) {
+          resolve(interceptedResponse);
+        } else {
+          resolve(interceptedResponse.body);
+        }
+      } catch (error) {
+        if (!(error instanceof HttpInterceptHaltError)) {
+          reject(error);
+        }
+      }
+    });
+  };
+
+  private createRequest(path: string, options?: HttpFetchOptions): Request {
+    const { query, prependBasePath: shouldPrependBasePath, ...fetchOptions } = merge(
+      {
+        method: 'GET',
+        credentials: 'same-origin',
+        prependBasePath: true,
+        headers: {
+          'kbn-version': this.params.kibanaVersion,
+          'Content-Type': 'application/json',
+        },
+      },
+      options || {}
```

Comment:
Shouldn't we delete all Kibana-specific options (`asResponse` as of now) to avoid a possible clash in the future?